File: src/sort/quick.py
#!/usr/bin/env python
# coding=utf-8

def quicksort(items, p, r):
    print(p, r, 'out')
    if p < r:
        q = partition(items, p, r)
        print(p, r, 'in', q)
        quicksort(items, p, q-1)
        quicksort(items, q+1, r)

def partition(items, p, r):
    x = items[r]
    i = p-1
    count = 0
    for j in range(p, r):
        if items[j] == x:
            count += 1

        if items[j] <= x:
            i = i + 1
            print(i, j, ' in for ')
            items[i],items[j] = items[j],items[i]

    items[i+1],items[r] = items[r],items[i+1]
    return int(i+1-count/2)



items = [2,5,9,3,7,0,-1]
print(items, 'before')
quicksort(items, 0, len(items)-1)
print(items)

File: src/sort/quick.js
function swap(arr, i, j) {
  let temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}
function quickSort(arr, p, r) {
  if (p < r) {
    const q = partition(arr, p, r);
    quickSort(arr, p, q - 1);
    quickSort(arr, q + 1, r);
  }
}

function partition(arr, p, r) {
  let i = p - 1;
  const x = arr[r];
  let count = 0;
  for(let j = p; j < r; j++) {
    if (arr[j] === x) {
      count += 1;
    }
    if (arr[j] <= x) {
      i += 1;
      swap(arr, i, j);
    }
  }
  swap(arr, i + 1, r);
  return Math.floor(i + 1 - count / 2);
}

const params = [1, 2, 6, 5, 7, 99, 10, 11, 9, 22, 14, 8];
console.log('sort before ', params);
quickSort(params, 0, params.length - 1);
console.log(' sorted ', params);